### PR TITLE
refactor(ApiClient): add hooks to handle side effects, handle users/me 

### PR DESCRIPTION
**Pull request by berdal84**
Added few little things in this branch:
- Auto connect when page loads (grabs the token from storage, and fetch user)
- Allows to signup
- Allow to logout (clear storage and user context)
- Separate ApiClient from components by introducing: useUserContext and useApiClient hooks. Those hooks handle API querying and maintaining a coherent user state.

### CODE REVIEW

File: src/hooks/useApiClient.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * Wraps ApiClient functions as a React hook.
+ * Each call may have a side effect on UserContext.
+ * 
+ * ReactComponent <----> useApiClient <-----> ApiClient <-----> API
+ *        ^                     v
+ *        |                     |
+ *        ----- UserContext <----
+ * 
+ * Condider bringing new functionnality here by calling ApiClient's functions.
```
Comment (rjmello) [inline src/hooks/useApiClient.tsx:15]:
```suggestion
 * Consider bringing new functionality here by calling ApiClient's functions.
```

File: src/hooks/useApiClient.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    console.debug("Automatic getme() success", _user)
+                    dispatch({ type: "loggedIn", username: _user.username })
+                }
+            }
+        }
+        fetch();
+    }, [])
+
+    /**
+     * SignIn using credentials and update user context accordingly
```
Comment (rjmello) [inline src/hooks/useApiClient.tsx:38]:
```suggestion
     * Sign up using credentials and update user context accordingly
```

File: src/hooks/useApiClient.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    const signup = useCallback( async (userCreate: UserCreate): Promise<boolean> => {
+        const success = await ApiClient.signup(userCreate);
+        if ( success ) {
+            dispatch({ type: 'loggedIn', username: userCreate.username})
+        } else {
+            dispatch({ type: 'error', message: "Unable to sign in"})
```
Comment (rjmello) [inline src/hooks/useApiClient.tsx:45]:
```suggestion
            dispatch({ type: 'error', message: "Unable to sign up"})
```

**Review by rjmello — CHANGES_REQUESTED** [review]

**Review by rjmello — APPROVED** [review]

**Comment by berdal84** [comment]
I just pushed a fix relative to the getOrCreateResource, was an error coming from the day I did the Axios refactor... Nothing complicated I'll merge directly.

**Review by rjmello — APPROVED** [review]

**Pull request merged**